scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,498 @@
+    /// Propagates errors.
+    pub async fn fetch_generation_by_timestamp(
+        &self,
+        time: &chrono::Duration,
+    ) -> Result<Option<GenerationTimestamp>, Box<dyn Error>> {
+        let query =
+            new_distributed_system_query(self.get_generation_by_timestamp_query(), &self.session)
+                .await?;
+
+        let result = self.session.query(query, (Timestamp(*time),)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            "
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time ASC
+    LIMIT 1;
+    ",
+            self.generations_table_name
+        )
+    }
+
+    /// Given a generation returns the next generation.
+    /// If given generation is currently operating, returns None.
+    /// Propagates errors.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &GenerationTimestamp,
+    ) -> Result<Option<GenerationTimestamp>, Box<dyn Error>> {
+        let query =
+            new_distributed_system_query(self.get_next_generation_query(), &self.session).await?;
+
+        let result = self.session.query(query, (generation,)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            "
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    ",
+            self.streams_table_name
+        )
+    }
+
+    /// Given a generation return identifiers of all streams of this generation.
+    pub async fn fetch_stream_ids(
+        &self,
+        generation: &GenerationTimestamp,
+    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
+        let mut result_vec = Vec::new();
+
+        let mut query =
+            new_distributed_system_query(self.get_stream_ids_by_time_query(), &self.session)
+                .await?;
+        query.set_page_size(DEFAULT_PAGE_SIZE);
+
+        let mut rows = self
+            .session
+            .query_iter(query, (generation,))
+            .await?
+            .into_typed::<(Vec<StreamID>,)>();
+
+        while let Some(next_row) = rows.next().await {
+            let (ids,) = next_row?;
+            for id in ids {
+                result_vec.push(id);
+            }
+        }
+        Ok(result_vec)
+    }
+
+    // Return single row containing generation.
+    fn return_single_row(
+        row: Option<Vec<Row>>,
+    ) -> Result<Option<GenerationTimestamp>, Box<dyn Error>> {
+        if let Some(row) = row {
+            if let Some(generation) = row.into_typed::<GenerationTimestamp>().next() {
+                return Ok(Some(generation?));
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+// Returns current cluster size in case of a success.
```

Comment:
Some of your functions are documented with `//` and some with `///`. Can we use `///` everywhere?

Reply:
I don't think thats a good idea since /// comments  are the ones that are parsed when creating docs.

Reply:
So your intention is to create doc comments only for public functions?

Reply:
Yes, I think it is a good idea to differentiate between comments intended to be read by a dev working on this repo or one that wants to know internal details by reading source code and ones intended for devs that want only to use exposed api.

Reply:
As you wish :)